service/kernel/data_management/train_test_split.cpp

```diff
@@ -0,0 +1,253 @@
+}
+
+template <typename DataType, typename IdxType>
+services::Status assignRows(const DataType * origDataPtr, const NumericTablePtr dataTable, const NumericTablePtr idxTable, const size_t startRow,
+                            const size_t nRows, const size_t nColumns)
+{
+    services::Status s;
+    BlockDescriptor<DataType> dataBlock;
+    BlockDescriptor<IdxType> idxBlock;
+    dataTable->getBlockOfRows(startRow, startRow + nRows, readWrite, dataBlock);
+    idxTable->getBlockOfColumnValues(0, startRow, nRows, readOnly, idxBlock);
+    DataType * dataPtr     = dataBlock.getBlockPtr();
+    const IdxType * idxPtr = idxBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(dataPtr);
+    DAAL_CHECK_MALLOC(idxPtr);
+
+    for (size_t i = 0; i < nRows; ++i)
+    {
+        const IdxType idx = idxPtr[i];
+        for (size_t j = 0; j < nColumns; ++j) dataPtr[i * nColumns + j] = origDataPtr[idxPtr[i] * nColumns + j];
+    }
+    return s;
+}
+
+template <typename DataType, typename IdxType, daal::CpuType cpu>
+services::Status assignRowsSubset(const DataType * origDataPtr, const NumericTablePtr dataTable, const NumericTablePtr idxTable, const size_t nRows,
+                                  const size_t nColumns, const size_t nThreads, const size_t blockSize)
+{
+    services::Status s;
+
+    if (nRows * nColumns > THREADING_BORDER && nThreads > 1)
+    {
+        const size_t nBlocks = nRows / blockSize + !!(nRows % blockSize);
+
+        daal::threader_for(nBlocks, nBlocks, [&](size_t iBlock) {
+            const size_t start = iBlock * blockSize;
+            const size_t end   = daal::services::internal::min<cpu, size_t>(start + blockSize, nRows);
+
+            s |= assignRows<DataType, IdxType>(origDataPtr, dataTable, idxTable, start, end - start, nColumns);
+        });
+    }
+    else
+    {
+        s |= assignRows<DataType, IdxType>(origDataPtr, dataTable, idxTable, 0, nRows, nColumns);
+    }
+
+    return s;
+}
+
+template <typename DataType, typename IdxType, daal::CpuType cpu>
+services::Status splitRows(const NumericTablePtr inputTable, const NumericTablePtr trainTable, const NumericTablePtr testTable,
+                           const NumericTablePtr trainIdxTable, const NumericTablePtr testIdxTable, const size_t nTrainRows, const size_t nTestRows,
+                           const size_t nColumns, const size_t nThreads)
+{
+    services::Status s;
+    const size_t blockSize = daal::services::internal::max<cpu, size_t>(BLOCK_CONST / nColumns, 1);
+    BlockDescriptor<DataType> origBlock;
+    inputTable->getBlockOfRows(0, nTrainRows + nTestRows, readOnly, origBlock);
+    const DataType * origDataPtr = origBlock.getBlockPtr();
+    DAAL_CHECK_MALLOC(origDataPtr);
+
+    s |= assignRowsSubset<DataType, IdxType, cpu>(origDataPtr, trainTable, trainIdxTable, nTrainRows, nColumns, nThreads, blockSize);
+    s |= assignRowsSubset<DataType, IdxType, cpu>(origDataPtr, testTable, testIdxTable, nTestRows, nColumns, nThreads, blockSize);
```

Comment:
You can add additional parallelism here. It will reduce threading overheads.

Reply:
I tried, it didn't affect performance somehow significantly.

Reply:
Ok, but it's strange for me.